lib/TPP/GPU/LinalgToXeGPU.cpp

```diff
@@ -0,0 +1,1394 @@
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/Dialect/XeGPU/IR/XeGPU.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+#include "llvm/ADT/TypeSwitch.h"
+
+#include <numeric>
+#include <optional>
+
+using namespace mlir;
+using namespace mlir::tpp;
+using namespace mlir::xegpu;
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_LINALGTOXEGPU
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+namespace {
+
+// Represents VNNI configuration for an operand.
+struct VnniConfig {
+  int vnniFactor;
+  int vnniAxis;
+};
+
+// Helper struct to keep track of tiles' position with respect to whole matrix.
+struct TilesArray {
+  TilesArray() = delete;
+  TilesArray(int numRows, int numCols) {
+    for (int i = 0; i < numRows; i++) {
+      tileMatrix.push_back(SmallVector<Value>{});
+      for (int j = 0; j < numCols; j++)
+        tileMatrix[i].push_back(Value{});
+    }
+  }
+  ~TilesArray() = default;
+
+  Value getTile(int row, int col) { return tileMatrix[row][col]; }
+
+  void setTile(int row, int col, Value val) { tileMatrix[row][col] = val; }
+
+  SmallVector<Value> toFlatVector() {
+    SmallVector<Value> flatVector;
+    for (auto row : tileMatrix)
+      flatVector.append(row);
+    return flatVector;
+  }
+
+  SmallVector<SmallVector<Value, 8>, 8> tileMatrix;
+};
+
+// Return DPAS tile sizes if the gemm-like operation fits DPAS hardware.
+static std::optional<SmallVector<int64_t>>
+getDPASConfig(linalg::LinalgOp linalgOp, int kTile) {
+  if (!(isa<linalg::MatmulOp>(linalgOp) ||
+        isa<linalg::BatchReduceMatmulOp>(linalgOp) ||
+        isa<linalg::GenericOp>(linalgOp))) {
+    return std::nullopt;
+  }
+
+  // Only static shapes are supported.
+  if (linalgOp.hasDynamicShape())
+    return std::nullopt;
+
+  auto aType = cast<ShapedType>(linalgOp.getDpsInputs()[0].getType());
+  auto bType = cast<ShapedType>(linalgOp.getDpsInputs()[1].getType());
+  auto cType = cast<ShapedType>(linalgOp.getDpsInits()[0].getType());
+
+  auto elemTypeA = aType.getElementType();
+  auto elemTypeB = bType.getElementType();
+  auto elemTypeC = cType.getElementType();
+
+  // TODO: Add more DPAS combinations.
+  bool isSupportedPrecision =
+      (elemTypeA.isF16() && elemTypeB.isF16() && elemTypeC.isF16()) ||
+      (elemTypeA.isF16() && elemTypeB.isF16() && elemTypeC.isF32());
+  if (!isSupportedPrecision)
+    return std::nullopt;
+
+  auto mDim = cType.getShape()[0];
+  auto nDim = cType.getShape()[1];
+  auto kDim = aType.getShape().back();
+
+  // DPAS hardware sizes in MxNxK format.
+  // TODO: In case more hardware configurations are available,
+  //       add some automatic selection for optimal sizes.
+  SmallVector<int64_t> dpasTile{8, 16, 16};
```

Comment:
this is the DPAS register block size for PVC. For ATS, it should be {8, 16, 8}.

Reply:
I'll expose it as a pass param for now. Later this could be pulled from target descriptor.

Reply:
Added pass param.